[1.16] Port/update Storage Bag

trunksbomb opened the pull request:
Features:

- 27 slot inventory
- controls for three different modes: Pickup, Deposit, and Refill
- Pickup mode controls if the bag will pickup nothing, everything, or just matching contents
- Deposit mode controls if the bag will deposit nothing, everything, or just matching contents into any inventory it's shift-clicked into
- Refill mode controls if the bag will refill the hotbar with matching contents (cobblestone in bag, cobblestone on bar -- keep cobblestone stacked to 64 if possible)
- Tooltip shows the modes it has

Commit:
begin porting Storage Bag from 1.12.
Implement registerClient() abstract function on ItemBase, so it can be called during Client registration to register item-based containers just like block-based containers already are
Bag is just an extended inventory at the moment, no extra features yet.

Commit:
Add sidebar to Storage Bag for buttons
Add ToggleButton class, easier to add buttons that hold "states" consisting of a button title, tooltip, and IPressable. Automatically moves between states on click.

Commit:
Implement a generalized Packet for sending ItemStackNBT data from client to server. Handles all INBT types except EndNBT (not sure what it is, don't think it would ever be sent though) and ListNBT (no easy way to construct/send that I saw).
add item pickup modes everything/filtered
add hotbar refill mode

Commit:
Prevent player from picking up any Storage Bag when its inventory is open
Finish implementing hit deposit dump/merge
Refactor translation components for item/button tooltip

trunksbomb commented:
![image](https://user-images.githubusercontent.com/1904331/101995717-3ebaed00-3c9a-11eb-8ec9-11ee9c6c6336.png)

trunksbomb commented:
![image](https://user-images.githubusercontent.com/1904331/101996093-4a5be300-3c9d-11eb-9767-b2b38d28f642.png)

increased to 81 slots on next commit. There's an item dupe right now though. Bag with pickup mode on everything/filtered will pick up the item into its own inventory but also put the same items in the player's inventory.

Lothrazar commented on src/main/java/com/lothrazar/cyclic/event/ItemEvents.java:
```diff
@@ -199,4 +199,31 @@ public void onHit(PlayerInteractEvent.LeftClickBlock event) {
   }
+
+  @SubscribeEvent
+  public void onPlayerPickup(EntityItemPickupEvent event) {
```
You mention a bug on pickup, where after the try insert into the bag, the player also gets a copy of the stack.  Would it be solved by cancelling the event when the insert is a success? either with .setCanceled(true); or .setResult(Result.DENY);

trunksbomb replied:
no I got it. I was just checking and comparing the wrong stacks after insertion. You don't want to cancel the event anyways, because you want the ItemEntity to go away if you took it all and you need to `ALLOW` the event for that.

Commit:
Add Patchouli guide entry
Adjust inventory size from 27 to 81 (same as Crate)
Add recipe
Fix item-dupe bug on pickup

Commit:
prevent putting bags in bags.
